File: JS/lab9/solutions/consumer.js
const { KafkaAvro } = require('kafkajs-avro')

const kafka = new KafkaAvro({
  brokers: ['localhost:9092', 'localhost:9093'],
  clientId: 'lab9',
  avro: {
    url: "http://localhost:8081"
  }
})

const topic = "disasters"
const consumer = kafka.avro.consumer({ groupId: `lab9-consumer` })

const run = async () => {
  await consumer.connect()
  await consumer.subscribe({ topic, fromBeginning: true })
  await consumer.run({
    eachMessage: async ({ topic, partition, message }) => {
      const prefix = `${topic}[${partition} | ${message.offset}] / ${message.timestamp}`
      console.log(`- ${prefix} - ${message.value}`)
    },
  })
}

run().catch(e => console.error(`[example/consumer] ${e.message}`, e))

File: Python/lab9/solutions/consumer.py

import io

from confluent_kafka import Consumer, KafkaError
from avro.io import DatumReader, BinaryDecoder
import avro.schema

schema = avro.schema.parse(open("DisasterValue.avsc").read())
reader = DatumReader(schema)

def decode(msg_value):
    message_bytes = io.BytesIO(msg_value)
    message_bytes.seek(5)
    decoder = BinaryDecoder(message_bytes)
    event_dict = reader.read(decoder)
    return event_dict

topic = "disasters"

consumer_config = {
    "group.id": 'disaster.example',
    "bootstrap.servers": 'localhost:9092',
}
c = Consumer(consumer_config)
c.subscribe([topic])
checks = 0
while checks < 1000:
    checks += 1
    msg = c.poll()
    if not msg.error():
        msg_value = msg.value()
        event_dict = decode(msg_value)
        print(event_dict)
    elif msg.error().code() != KafkaError._PARTITION_EOF:
        print(msg.error())
        running = False
